PLUGIN-63 Hubspot (SaaS source) batchsource

ArturLytvynenko opened the pull request:
PLUGIN-63 Hubspot (SaaS source) batchsource
https://issues.cask.co/browse/PLUGIN-63
https://wiki.cask.co/display/CE/Hubspot+plugins
in this PR:
Hubspot batch source

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batchsource

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+
+Description
+-----------
+HubSpot is a developer and marketer of software products for inbound marketing and sales. HubSpot is an inbound marketing and sales platform that helps companies attract visitors, convert leads, and close customers. Its products and services aim to provide tools for social media marketing, content management, web analytics, CRM and search engine optimization. They expose an API that allows you to ...
```
Description should concentrate more on plugin specifics. Not what HubSpot is. I think users of this plugin will know what HuSpot is :) You may add some very brief info on what HubSpot is though.

"The plugins will allow users to pull ..." -> "The plugins allows to pull ..."

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+
+**Object(s) to pull:** Select from Contact Lists, Contacts, Email Events, Email Subscription, Recent Campaigns, Analytics, Companies, Deals, Deal Pipelines, Marketing Email, Products, Tickets
+
+**Number of API calls per day:** The number of API calls to make per day, to avoid hitting HubSpot's rate limits
```
Don't start the descriptions with "the, a, an". This was mentioned by Albert to me.

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+
+**API Key:** OAuth2 API Key.
+
+**Object(s) to pull:** Select from Contact Lists, Contacts, Email Events, Email Subscription, Recent Campaigns, Analytics, Companies, Deals, Deal Pipelines, Marketing Email, Products, Tickets
```
The description usually starts with a noun and answer the question "what is this property?"

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+### Analytics
+
+**Time Period:** The time period used to group the data. Must be one of:
+* total - Data is rolled up to totals covering the specified time.
```
it's a bit not clear what this means.

ArturLytvynenko replied:
description from official documentation

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+* summarize/weekly - Grouped by week, data is rolled up across all breakdowns
+* summarize/monthly - Grouped by month, data is rolled up across all breakdowns
+
+NOTE: When using daily, weekly, or monthly for the time_period, you must include at least one filter 
```
for Time Period.

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+* knowledge-articles - Pull data for knowledge base articles.
+
+The analytics report category used to break down the analytics data. Must be one of:
+* totals - Data will be the totals rolled up from
```
it's a bit not clear what this means.

ArturLytvynenko replied:
description from official documentation

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+The analytics report category used to break down the analytics data. Must be one of:
+* totals - Data will be the totals rolled up from
+* sessions - Data broken down by session details
+* sources - Data broken down by traffic source
```
these all lines should end with a dot probably?

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+
+**End Date:** End date for the report data. YYYYMMDD format.
+
+**Filters:** Filter the analytics report data to include only the specified breakdowns.
```
The description usually starts with a noun and answer the question "what is this property?"

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/batch/source/HubspotBatchSource.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.stream.Collectors;
+
+/**
+ * Plugin read Hubspot objects in batch
```
Plugin reads

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/batch/source/HubspotBatchSource.java:
```diff
@@ -0,0 +1,82 @@
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(HubspotBatchSource.NAME)
+@Description("Plugin read Hubspot objects in batch")
```
same

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/batch/source/HubspotBatchSource.java:
```diff
@@ -0,0 +1,82 @@
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    IdUtils.validateId(config.referenceName);
```
`IdUtils.validateReferenceName(config.referenceName, failureCollector)` and probably move to validate method

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/batch/source/HubspotInputFormat.java:
```diff
@@ -0,0 +1,42 @@
+
+  @Override
+  public List<InputSplit> getSplits(JobContext jobContext) {
+    return Collections.singletonList(new HubspotSplit());
```
We should try to parallalize the actions anywhere it's possible. Can you please state the reasons. Why here we cannot parallelize or add parallelization code?

ArturLytvynenko replied:
Some pages has offset value as key string,
so to get next page, need to pull previous

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/batch/source/HubspotInputFormatProvider.java:
```diff
@@ -0,0 +1,50 @@
+  public static final String PROPERTY_CONFIG_JSON = "cdap.hubspot.config";
+  private static final Gson gson = new GsonBuilder().create();
+  private final Map<String, String> conf;
+
```
nit: remove empty line

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+  public static final String END_DATE = "endDate";
+  public static final String FILTERS = "filters";
+  @Name(FILTERS)
+  @Description(" Filter the analytics report data to include only the specified breakdowns.")
```
remove leading space.

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+/**
+ * Provides base configuration for accessing Hubspot API
+ */
+public class BaseHubspotConfig extends ReferencePluginConfig {
```
Please update the descriptions accordingly to the doc.

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+    try {
+      getTimePeriod();
+    } catch (IllegalArgumentException e) {
+      failureCollector.addFailure(String.format("Time Period '%s' is not a valid Time Period", timePeriod),
```
`Time period '%s' is not a valid.`,
also add corrective action with the correct format.

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+  }
+
+  private void validateFilters(FailureCollector failureCollector) {
+    if (containsMacro(FILTERS) &&
```
`||`?

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+    }
+    List<String> filters = getFilters();
+    for (String filter : filters) {
+      if (filter.isEmpty()) {
```
This is strange validation. Only for empty value. This is what you intended?

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+    List<String> filters = getFilters();
+    for (String filter : filters) {
+      if (filter.isEmpty()) {
+        failureCollector.addFailure(String.format("Filter '%s' is not a valid filter", filter),
```
ll the failure messages and corrective actions should end with '.'

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+    try {
+      getReportType();
+    } catch (IllegalArgumentException e) {
+      failureCollector.addFailure(String.format("reportFormat '%s' is not a valid Report Type", reportType),
```
- use Display name not variable name. 
- ends with dot.

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+      case WEEKLY:
+      case MONTHLY:
+        if (filters.isEmpty()) {
+          failureCollector.addFailure("When using daily, weekly, or monthly for the time_period," +
```
- use Display name not variable name. 
- ends with dot.

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+    try {
+      getObjectType();
+    } catch (IllegalArgumentException e) {
+      failureCollector.addFailure(String.format("reportFormat '%s' is not a valid Object Type", objectType),
```
same

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+      || containsMacro(APP_ID)) {
+      return;
+    }
+    //todo validate
```
Fix this. We should not present a code to google with any TODOs.

dgrinenko replied:
and? do we still need this?

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    httpClientBuilder.setDefaultHeaders(clientHeaders);
+    CloseableHttpClient client = httpClientBuilder.build();
+
+    int maxTries = 3;
```
private static final

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    while (true) {
+      CloseableHttpResponse response = client.execute(request);
+      StatusLine statusLine = response.getStatusLine();
+      if (statusLine.getStatusCode() != 200) {
```
What about 201 202?
`if (statusLine.getStatusCode() >= 300) {`

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+  public static final String START_DATE = "startDate";
+  public static final String END_DATE = "endDate";
+  public static final String FILTERS = "filters";
+  @Name(FILTERS)
```
i guess, static constants need to be separated with one blank line from class fields

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+        b.addParameter("f", filter);
+      }
+      if (getLimitPropertyName(baseHubspotConfig) != null) {
+        b.addParameter(getLimitPropertyName(baseHubspotConfig), "100");
```
No magic numbers should be in code. `private static final` with a good descriptive name. and a comment what is this and why we use this value.

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+  public String endDate;
+  @Name(REPORT_TYPE)
+  @Description("The analytics report type of content that you want to get data for. Must be one of:\n" +
+    "\n" +
```
it is too long for the pop-up description, same as for the annotation. Please make a short description, while the rest should be included to the doc

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+  @Nullable
+  public String reportType;
+  @Name(TIME_PERIOD)
+  @Description("The time period used to group the data. Must be one of:\n" +
```
make it in one sentence, generic description

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+      }
+      uri = b.build();
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(e);
```
add a message along with causing exception.

aonischuk commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    JsonObject jsonObject = root.getAsJsonObject();
+    List<JsonElement> hubspotObjects = new ArrayList<>();
+    if (getObjectApiName(baseHubspotConfig) != null) {
+      JsonArray jsonObjects = jsonObject.get(getObjectApiName(baseHubspotConfig)).getAsJsonArray();
```
What if this is not array? Can we give an informative error message?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+  @Macro
+  public String apiKey;
+  @Name(OBJECT_TYPE)
+  @Description("Select from Contact Lists, Contacts, Email Events, Email Subscription," +
```
no need to list here all objects, just replace them with like "select from Hubspot objects/fields. i.e. any what fits by the logic"

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/BaseHubspotConfig.java:
```diff
@@ -0,0 +1,266 @@
+    }
+  }
+
+  private void validateTimePeriod(FailureCollector failureCollector) {
```
can be all validation be moved to external static validation class?  It's not really good to mix Configuration POJO class and some validation logic.

aonischuk commented on widgets/HubspotBatchSource-batchsource.json:
```diff
@@ -0,0 +1,129 @@
+        {
+          "widget-type": "textbox",
+          "label": "App ID",
+          "name": "appId"
```
is this secret or not? If it is should be password type.

aonischuk commented on widgets/HubspotBatchSource-batchsource.json:
```diff
@@ -0,0 +1,129 @@
+        {
+          "widget-type": "textbox",
+          "label": "API Key",
+          "name": "apiKey"
```
is this secret or not? If it is should be password type.

dgrinenko replied:
it is definitely a secret, the type should be `securekey-text`

aonischuk commented on widgets/HubspotBatchSource-batchsource.json:
```diff
@@ -0,0 +1,129 @@
+        {
+          "widget-type": "textbox",
+          "label": "Start Date",
+          "name": "startDate"
```
add placeholder with format.

aonischuk commented on widgets/HubspotBatchSource-batchsource.json:
```diff
@@ -0,0 +1,129 @@
+        {
+          "widget-type": "textbox",
+          "label": "End Date",
+          "name": "endDate"
```
add placeholder with format.

aonischuk commented on docs/HubspotBatchSource-batchsource.md:
```diff
@@ -0,0 +1,56 @@
+# Google Ads - batch source
```
I think you forgot to include pom.xml

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    return new HubspotPage(hubspotObjects, baseHubspotConfig, offset, hasNext);
+  }
+
+  private String getLimitPropertyName(BaseHubspotConfig baseHubspotConfig) {
```
Should be this annotated  as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    }
+  }
+
+  private String getOffsetPropertyName(BaseHubspotConfig baseHubspotConfig) {
```
should be this annotated as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    }
+  }
+
+  private String getOffsetApiName(BaseHubspotConfig baseHubspotConfig) {
```
should be this annotated as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    }
+  }
+
+  private String getMoreApiName(BaseHubspotConfig baseHubspotConfig) {
```
should be this annotated as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    }
+  }
+
+  private String getObjectApiName(BaseHubspotConfig baseHubspotConfig) {
```
should be this annotated as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java:
```diff
@@ -0,0 +1,288 @@
+    }
+  }
+
+  public String getEndpoint(BaseHubspotConfig baseHubspotConfig) {
```
should be this annotated as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotPage.java:
```diff
@@ -0,0 +1,54 @@
+    return hubspotObjects.iterator();
+  }
+
+  public HubspotPage nextPage() throws IOException {
```
should be this annotated as @nullable?

dgrinenko commented on src/main/java/io/cdap/plugin/hubspot/common/HubspotPage.java:
```diff
@@ -0,0 +1,54 @@
+      return new HubspotHelper().getHupspotPage(hubspotConfig, offset);
+    } else {
+      return null;
+    }
```
return (hasNext)
 ? new HubspotHelper().getHupspotPage(hubspotConfig, offset)
 : null;

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batchsource

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+  <version>1.0.0-SNAPSHOT</version>
+
+  <properties>
+    <awaitility.version>3.1.6</awaitility.version>
```
This seems like not used. Please check and remove any deps that are not used.

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+    <hydrator.version>2.3.0-SNAPSHOT</hydrator.version>
+    <jackson.version>2.9.9</jackson.version>
+    <junit.version>4.11</junit.version>
+    <jython.version>2.7.1</jython.version>
```
seems like we don't use this.

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+    <jackson.version>2.9.9</jackson.version>
+    <junit.version>4.11</junit.version>
+    <jython.version>2.7.1</jython.version>
+    <mockito.version>1.10.19</mockito.version>
```
seems like we don't use this.

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+    <jython.version>2.7.1</jython.version>
+    <mockito.version>1.10.19</mockito.version>
+    <spark2.version>2.1.3</spark2.version>
+    <unxml.version>0.9</unxml.version>
```
seems like we don't use this.

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+    <junit.version>4.11</junit.version>
+    <jython.version>2.7.1</jython.version>
+    <mockito.version>1.10.19</mockito.version>
+    <spark2.version>2.1.3</spark2.version>
```
seems like we don't use this.

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+    <mockito.version>1.10.19</mockito.version>
+    <spark2.version>2.1.3</spark2.version>
+    <unxml.version>0.9</unxml.version>
+    <wiremock.version>1.49</wiremock.version>
```
seems like we don't use this.

aonischuk commented on pom.xml:
```diff
@@ -0,0 +1,557 @@
+          <dependency>
+            <groupId>com.puppycrawl.tools</groupId>
+            <artifactId>checkstyle</artifactId>
+            <version>6.19</version>
```
8.18

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batch source

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batch source

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batch source

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source

doc changes

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source tests

PLUGIN-63 Hubspot (SaaS source) batch source

testing

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batch source

ArturLytvynenko committed:
PLUGIN-63 Hubspot (SaaS source) batch source
